ensime_shared/editor.py

```diff
@@ -0,0 +1,384 @@
+        """Get a line from the current buffer.
+
+        Args:
+            lnum (Optional[str]): Number of the line to get, current if ``None``.
+
+        Todo:
+            - Give this more behavior of Vim ``getline()``?
+            - ``buffer[index]`` is zero-based, this is probably too confusing
+        """
+        return self._vim.current.buffer[lnum] if lnum else self._vim.current.line
+
+    def getlines(self, bufnr=None):
+        """Get all lines of a buffer as a list.
+
+        Args:
+            bufnr (Optional[int]): A Vim buffer number, current if ``None``.
+
+        Returns:
+            List[str]
+        """
+        buf = self._vim.buffers[bufnr] if bufnr else self._vim.current.buffer
+        return buf[:]
+
+    def goto(self, offset):
+        """Go to a specific byte offset in the current buffer."""
+        self._vim.command('goto {}'.format(offset))
+
+    def menu(self, prompt, choices):
+        """Presents a selection menu and returns the user's choice.
+
+        Args:
+            prompt (str): Text to ask the user what to select.
+            choices (Sequence[str]): Values for the user to select from.
+
+        Returns:
+            The value selected by the user, or ``None``.
+
+        Todo:
+            Nice opportunity to provide a hook for Unite.vim, etc. here.
+        """
+        menu = [prompt] + [
+            "{0}. {1}".format(*choice) for choice in enumerate(choices, start=1)
+        ]
+        command = 'inputlist({})'.format(repr(menu))
+        choice = int(self._vim.eval(command))
+
+        # Vim returns weird stuff if user clicks outside choices with mouse
+        if not 0 < choice < len(menu):
+            return
+
+        return choices[choice - 1]
+
+    def isneovim(self):
+        """Whether or not the underlying editor is Neovim. Use this sparingly.
+
+        :returns: TODO
+        """
+        # TODO: could be memoized
+        return bool(self._vim.eval("has('nvim')"))
+
+    def set_buffer_options(self, options, bufnr=None):
+        """Set buffer-local options for a buffer, defaulting to current.
+
+        Args:
+            options (dict):
+                Options to set, with keys being Vim option names. For Boolean
+                options, use a :class:`bool` value as expected, e.g.
+                ``{'buflisted': False}`` for ``setlocal nobuflisted``.
+            bufnr (Optional[int]):
+                A Vim buffer number, as you might get from VimL ``bufnr('%')``
+                or Python ``vim.current.buffer.number``. If ``None``, options
+                are set on the current buffer.
+        """
+        buf = self._vim.buffers[bufnr] if bufnr else self._vim.current.buffer
+
+        for opt, value in options.items():
+            buf.options[opt] = value
+
+    # TODO: make this a R/W property?
+    def set_filetype(self, filetype):
+        """Set filetype for the current buffer.
+
+        Unfortunately setting the filetype with ``set_buffer_options`` (or
```

Comment:
This is an implementation detail, no need to be leaked in the editor API. Maybe it could be specially handled in `set_buffer_options` instead of being a separate method.